src/main/java/fr/xephi/authme/service/GeoIpService.java

```diff
@@ -49,94 +83,173 @@
-        if (downloadTask != null && downloadTask.isAlive()) {
+        if (downloading) {
+            // we are currently downloading the database
             return false;
         }
-        if (lookupService != null) {
+
+        if (databaseReader != null) {
+            // everything is initialized
             return true;
         }
 
-        if (dataFile.exists()) {
-            boolean dataIsOld = (System.currentTimeMillis() - dataFile.lastModified()) > TimeUnit.DAYS.toMillis(30);
-            if (!dataIsOld) {
-                try {
-                    lookupService = new LookupService(dataFile, GEOIP_MEMORY_CACHE);
+        if (Files.exists(dataFile)) {
+            try {
+                FileTime lastModifiedTime = Files.getLastModifiedTime(dataFile);
+                if (Duration.between(lastModifiedTime.toInstant(), Instant.now()).toDays() <= UPDATE_INTERVAL) {
+                    databaseReader = new Reader(dataFile.toFile(), FileMode.MEMORY, new CHMCache());
                     ConsoleLogger.info(LICENSE);
+
+                    // don't fire the update task - we are up to date
                     return true;
-                } catch (IOException e) {
-                    ConsoleLogger.logException("Failed to load GeoLiteAPI database", e);
-                    return false;
                 }
-            } else {
-                FileUtils.delete(dataFile);
+            } catch (IOException ioEx) {
+                ConsoleLogger.logException("Failed to load GeoLiteAPI database", ioEx);
+                return false;
             }
         }
-        // Ok, let's try to download the data file!
-        downloadTask = createDownloadTask();
-        downloadTask.start();
+
+        // File is outdated or doesn't exist - let's try to download the data file!
+        startDownloadTask();
         return false;
     }
 
     /**
      * Create a thread which will attempt to download new data from the GeoLite website.
+     */
+    private void startDownloadTask() {
+        downloading = true;
+
+        // use bukkit's cached threads
+        bukkitService.runTaskAsynchronously(() -> {
+            try {
+                // download database to temporarily location
+                Path tempFile = Files.createTempFile(ARCHIVE_FILE, null);
+                try (OutputStream out = Files.newOutputStream(tempFile)) {
+                    Resources.copy(new URL(ARCHIVE_URL), out);
+                }
+
+                // MD5 checksum verification
+                String targetChecksum = Resources.toString(new URL(CHECKSUM_URL), StandardCharsets.UTF_8);
+                if (!verifyChecksum(Hashing.md5(), tempFile, targetChecksum)) {
+                    return;
+                }
+
+                // tar extract database and copy to target destination
+                if (!extractDatabase(tempFile, dataFile)) {
+                    ConsoleLogger.warning("Cannot find database inside downloaded GEO IP file at " + tempFile);
```

Comment:
Should we `return;` after this or should `downloading` really be set to `false` in this case?

Reply:
Thanks for pointing out. `return` is missing here.